src/lib/run-dev.js

```diff
@@ -0,0 +1,165 @@
+const bundle = require('./bundle')
+const serve = require('./serve')
+const Cleanup = require('./cleanup')
+const runDevLocal = require('./run-dev-local')
+
+const buildActions = require('./build-actions')
+const deployActions = require('./deploy-actions')
+const actionsWatcher = require('./actions-watcher')
+
+const utils = require('./app-helper')
+const execa = require('execa')
+const { run: logPoller } = require('./log-poller')
+
+/** @private */
+async function runDev (args = [], config, options = {}, log = () => {}) {
+  // note: args are considered perfect here because this function is only ever called by the `app run` command
+
+  /* parcel bundle options */
+  const bundleOptions = options.parcel || {}
+  /* skip actions */
+  const skipActions = !!options.skipActions
+  /* fetch logs for actions option */
+  const fetchLogs = options.fetchLogs || false
+
+  // control variables
+  const hasFrontend = config.app.hasFrontend
+  const withBackend = config.app.hasBackend && !skipActions
+  const isLocal = !options.devRemote // applies only for backend
+
+  aioLogger.debug(`hasFrontend ${hasFrontend}`)
+  aioLogger.debug(`withBackend ${withBackend}`)
+  aioLogger.debug(`isLocal ${isLocal}`)
+
+  let frontEndUrl
+
+  // state
+  let devConfig = config // config will be different if local or remote
+  devConfig.envFile = '.env'
+
+  const cleanup = new Cleanup()
+
+  // bind cleanup function
+  process.on('SIGINT', async () => {
+    // in case app-scripts are eventually turned into a lib:
+    // - don't exit the process, just make sure we get out of waiting
+    // - unregister sigint and return properly (e.g. not waiting on stdin.resume anymore)
+    try {
+      await cleanup.run()
+      log('exiting!')
+      process.exit(0) // eslint-disable-line no-process-exit
+    } catch (e) {
+      aioLogger.error('unexpected error while cleaning up!')
+      aioLogger.error(e)
+      process.exit(1) // eslint-disable-line no-process-exit
+    }
+  })
+
+  try {
+    // TODO: Is there a chance of using run cmd on actions from plugin-runtime in future?
+    // In that case this backend stuff might have to go to lib-runtime ?
+
+    // Build Phase
+    if (withBackend) {
+      if (isLocal) {
+        const { config: localConfig, cleanup: localCleanup } = await runDevLocal(config, log, options.verbose)
+        devConfig = localConfig
+        cleanup.add(() => localCleanup(), 'cleaning up runDevLocal')
+      } else {
+        // check credentials
+        rtLibUtils.checkOpenWhiskCredentials(devConfig)
+        log('using remote actions')
+      }
+
+      // build and deploy actions
+      log('building actions..')
+      await buildActions(devConfig)
+
+      log(`watching action files at ${devConfig.actions.src}...`)
+      const { cleanup: watcherCleanup } = await actionsWatcher({ config: devConfig, isLocal, log })
+      cleanup.add(() => watcherCleanup(), 'stopping action watcher...')
+    }
+
+    if (hasFrontend) {
```

Comment:
This is confusing with 2 if(hasFrontend) cases, I know that they are both doing different things at different stages, but I think this section needs some more comments to explain what/why this is the way it is.

Reply:
fixed in 63d4e58